openmeter/ent/schema/billing.go

```diff
@@ -0,0 +1,418 @@
+
+		field.Enum("invoice_line_item_resolution").
+			GoType(billing.GranualityResolution("")),
+
+		field.Bool("invoice_line_item_per_subject").
+			Default(false),
+	}
+}
+
+func (BillingWorkflowConfig) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+	}
+}
+
+func (BillingWorkflowConfig) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("billing_invoices", BillingInvoice.Type),
+		edge.To("billing_profile", BillingProfile.Type),
+	}
+}
+
+type BillingInvoiceItem struct {
+	ent.Schema
+}
+
+func (BillingInvoiceItem) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (BillingInvoiceItem) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("invoice_id").
+			Optional().
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
+			}),
+		field.String("customer_id").
+			NotEmpty().
+			Immutable().
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
+			}),
+
+		field.Time("period_start"),
+		field.Time("period_end"),
+		field.Time("invoice_at"),
+
+		// TODO[dependency]: overrides (as soon as plan override entities are ready)
+
+		field.Other("quantity", alpacadecimal.Decimal{}).
+			SchemaType(map[string]string{
+				"postgres": "numeric",
+			}),
+		field.Other("unit_price", alpacadecimal.Decimal{}).
+			SchemaType(map[string]string{
+				"postgres": "numeric",
+			}),
+		field.String("currency").
+			NotEmpty().
+			Immutable().
+			SchemaType(map[string]string{
+				"postgres": "varchar(3)",
+			}),
+		field.JSON("tax_code_override", invoice.TaxOverrides{}).
+			SchemaType(map[string]string{
+				"postgres": "jsonb",
+			}),
+	}
+}
+
+func (BillingInvoiceItem) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "invoice_id"),
+		index.Fields("namespace", "customer_id"),
+	}
+}
+
+func (BillingInvoiceItem) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("billing_invoice", BillingInvoice.Type).
+			Ref("billing_invoice_items").
+			Field("invoice_id").
+			Unique(),
+		// TODO[dependency]: Customer edge, as soon as customer entities are ready
+
+	}
+}
+
+type BillingInvoice struct {
+	ent.Schema
+}
+
+func (BillingInvoice) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (BillingInvoice) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("key").
+			NotEmpty().
+			Immutable(),
+		field.String("customer_id").
+			NotEmpty().
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
+			}).
+			Immutable(),
+		field.String("billing_profile_id").
+			NotEmpty().
+			Immutable().
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
+			}),
+		field.Time("voided_at").
+			Optional(),
+		field.String("currency").
```

Comment:
we will use lot's of currencies. Mixin?

Reply:
My concern with mixins for this is that in case of plans, I don't know if we would want to make that field immutable. With mixins you cannot fine-tune these things (e.g. change the currency in a draft plan).